public/app/features/query/components/QueryEditorRow.tsx

```diff
@@ -273,18 +280,32 @@ export class QueryEditorRow<TQuery extends DataQuery> extends PureComponent<Prop
   };
 
   onRemoveQuery = () => {
-    this.props.onRemoveQuery(this.props.query);
+    const { onRemoveQuery, query, trackActions } = this.props;
+    onRemoveQuery(query);
+
+    if (trackActions) {
+      reportInteraction(trackActions.remove);
+    }
   };
 
   onCopyQuery = () => {
-    const copy = cloneDeep(this.props.query);
-    this.props.onAddQuery(copy);
+    const { query, onAddQuery, trackActions } = this.props;
+    const copy = cloneDeep(query);
+    onAddQuery(copy);
+
+    if (trackActions) {
+      reportInteraction(trackActions.duplicateQuery);
+    }
   };
 
   onDisableQuery = () => {
-    const { query } = this.props;
-    this.props.onChange({ ...query, hide: !query.hide });
-    this.props.onRunQuery();
+    const { query, onChange, onRunQuery, trackActions } = this.props;
+    onChange({ ...query, hide: !query.hide });
+    onRunQuery();
+
+    if (trackActions) {
+      reportInteraction(trackActions.disableEnableQuery);
+    }
```

Comment:
i'm wondering wether this belongs here. This component shouldn't care at all about tracking, and maybe we should bubble up all those events so that they can be handled on the consumer side, which can then add tracking.

Reply:
can you please explain how adding a standard`onQueryCopied` or something similar (which this component already partially has, and is widely used) is more confusing than a trackingContext property that when set makes the component track interaction with hardcoded names and values?

Reply:
Just to get it right let me rephrase what I understood from your comments. Please correct me if I got it wrong:
- create 3 functions for tracking in _public/app/features/explore/**QueryRows.tsx**_ for `onQueryCopied`, `onQueryRemoved` and `onQueryEnabled`
- pass them through _public/app/features/query/components/**QueryEditorRows.tsx**_ to _public/app/features/query/components/**QueryEditorRow.tsx**_ so they can get called when one of the 3 buttons is clicked
- So every time someone wants to implement tracking for their use case in _public/app/features/query/components/**QueryEditorRow.tsx**_  they have to create their own functions because we would like to keep the component in grafana-ui as general as possible

This sounds unnecessary complex and there is plenty of code that could not be reused. I am not questioning the practice of passing through callbacks but in this case it seems too much.
Currently the component in Explore is passing a simple string to the ui component. It is obvious where it is being used and for which use case. To make it even more obvious I could leave a comment what is expected in the string.

I implemented the tracking in a component that is widely used because there are plenty of teams working on tracking and they can reuse it.

What do you think?

Reply:
i think this component responsibility does not include tracking, and it should be implemented separately by its consumers.
it may seem obvious to you how this component works, but it may not be to everyone. what you are proposing here is not done anywhere else.

Reply:
There's already tracking in `QueryEditorRows`. `QueryEditorRow` is just an internal component used by "rows" container. That means it will be consistent with the current approach to add tracking to `QueryRow` (not sure if it's the best way but at least we should be consistent so we can change it later more easily). 

The concerns you both have are valid and it all depends on how we treat `QueryEditorRow(s)` component. If we ever plan to make it generic as `grafana-ui` components it cannot have a dependency to `reportInteractions` (it's part of the runtime and it would make it not reusable), but that would also require moving out other events. If the purpose of this component is to be used only in core Grafana we should probably:

* make names consistent - e.g. we have `query_row_reorder_ended` and `grafana_queryRow_remove_query` 
* `trackingContext` should also be added to other dragging events, and possibly to all consumers of QueryEditorRow
* unless we want to track these actions regardless of "context", then we can add tracking without it.

At the moment what is probably confusing is that we limit some interaction tracking to Explore only.

Reply:
IMO, we should be careful when adding tracking to reusable components. They can break other people's metrics if you use them as part of a funnel since you can start getting more traffic that doesn't belong to your experience. 

I suggest not reporting interaction from the component itself but doing it via callback from the upper level that is used only once and it belongs only to a product.

Reply:
About naming, using grafana as the namespace is too vague, I'd suggest using: `grafana_explore` as a prefix to identify easily all the product events as you do in the other events. WDYT?